### ISSUE TITLE
DETR tutorials to use it on custom data

### ISSUE THREAD

**Issue by NielsRogge** [facebookresearch/detr#428]
Some months ago, I've added DETR to [HuggingFace Transformers](https://github.com/huggingface/transformers). I've replaced the original torchvision backbones such that you can use any backbone available in the [timm](https://github.com/rwightman/pytorch-image-models) repository (like EfficientNets or MobileNets etc.). The model is implemented using the same API as other models in HuggingFace like BERT (i.e. you have `DetrModel` which is the encoder-decoder Transformer without any head on top, `DetrForObjectDetection` which has the object detection and class labels classifier heads on top and `DetrForSegmentation`, which adds the mask head on top). 

The model weights are hosted on the HuggingFace [hub](https://huggingface.co/models?search=detr).

I've made 5 different notebooks, illustrating how to use DETR both for inference and training on custom data, both for object detection and panoptic segmentation. You can find them here: https://github.com/NielsRogge/Transformers-Tutorials

Hope it helps for people to easily use DETR!

**Comment by ducvuuit** [comment]
![image](https://user-images.githubusercontent.com/62535444/131203061-4349799f-d760-49e3-b9f7-0fe0333e1920.png)

**Comment by ducvuuit** [comment]
![image](https://user-images.githubusercontent.com/62535444/131203370-e0405fae-2f6b-4a76-a188-819903b68f94.png)
I train on my dataset, at epoch 0, it always stops at 5% without any error, Ram and memory are always guaranteed during training.
Does anyone have the same problem as me?

**Comment by NielsRogge** [comment]
Hi, I've just run my notebook, it still works fine. That's weird. You can perhaps restart the runtime and run again?

**Comment by ducvuuit** [comment]
Hi @NielsRogge, I fixed the error above, only readjust "max_steps". But the time for training is too long. With 1 epoch takes over **7 hours** while the official source code only takes about **30 minutes** for each epoch. I adjusted the batch size but also take over 7 hours. How to solve this problem?
Thank you in advance.

**Comment by NielsRogge** [comment]
Yes, it takes 30 minutes per epoch if you have 8 GPUs, as stated in the README. On a single GPU, it will take a bit longer. ;)

**Comment by mytk2012** [comment]
It seems that authors advise people to finetune model to apply in custom dataset. If we have enough datasets(>10k), shall we need to fintune it? It's weired we must finetune detr model , but CNN model can be applied in coco-format without finetuing.

**Comment by ver0z** [comment]
Do you know if it is possible to use DeepSpeed with DETR ? It could help to speedup the training, don't it ?

**Comment by NielsRogge** [comment]
Hi,

DeepSpeed is mostly meant to fit very big models in one or more GPUs. Its use case is not to speed up training, iirc.

**Comment by ver0z** [comment]
So if I pass a big number of batch could it be useful in this case ?

**Comment by ohhenrylee** [comment]
During training, is the CNN backbone trained at the same time as DETR or will pre-training be required?

**Comment by NielsRogge** [comment]
The CNN is trained at the same time as the encoder-decoder Transformer, however one starts with a pre-trained CNN and a randomly initialized encoder-decoder Transformer. One also typically uses a different learning rate between the CNN backbone and the Transformer.

**Comment by VerrenRatanaP** [comment]
Hi @NielsRogge is there a tutorial on changing the backbone of DETR?
I've been trying to change the backbone to Vision Transformer in the backbone.py but I encountered a number of errors. 
I've used #154 as my reference. 

Another approach that I have tried is to use your [tutorial](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/DETR/Fine_tuning_DetrForObjectDetection_on_custom_dataset_(balloon).ipynb) and replaced `DetrImageProcessor.from_pretrained("facebook/detr-resnet-50")` with
`ViTFeatureExtractor.from_pretrained('google/vit-base-patch16-224-in21k')`
However, I'm not sure if this is the right implementation. Could you advise?

**Comment by NielsRogge** [comment]
Hi,

Replacing the backbone can be done as follows:
```
from transformers import ConvNextConfig, DetrConfig, DetrForObjectDetection

backbone_config = ConvNextConfig(out_features=["stage1", "stage2", "stage3", "stage4"])
config = DetrConfig(backbone_config=backbone_config, use_timm_backbone=False)

model = DetrForObjectDetection(config)
```
This will work out-of-the-box for convolutional backbones like ConvNeXt:

```
import torch

pixel_values = torch.randn(1, 3, 224, 224)
outputs = model(pixel_values)
```

 For a vision transformer-based backbone, I'd recommend using ViTDet:
```
from transformers import VitDetConfig, DetrConfig, DetrForObjectDetection

backbone_config = VitDetConfig(out_features=["stage1", "stage2", "stage3", "stage4"])
config = DetrConfig(backbone_config=backbone_config,  use_timm_backbone=False)

model = DetrForObjectDetection(config)
```

**Comment by truong2710-cyber** [comment]
Hi @NielsRogge, I have a question.
If I trained a DeTR model with your notebook (in HuggingFace), can I convert the checkpoint back to the format of the original DeTR repo ([https://github.com/facebookresearch/detr])?
Thanks.

**Comment by truong2710-cyber** [comment]
Btw, I have a task in which I have to train Detr on remapped COCO dataset containing only 2 classes (class 1: person, class 2: vehicle corresponding to 3 classes in original COCO (car, truck, bus)). What should I do?

**Comment by truong2710-cyber** [comment]
Thanks a lot @NielsRogge . If I have a A100 GPU, could you estimate how long does it take to finetune the model on my remapped COCO dataset?

**Comment by truong2710-cyber** [comment]
> > Btw, I have a task in which I have to train Detr on remapped COCO dataset containing only 2 classes (class 1: person, class 2: vehicle corresponding to 3 classes in original COCO (car, truck, bus)). What should I do?
> 
> Have you managed to convert a COCO dataset to Huggingface dataset format? If so, could share the steps?

I think Huggingface directly uses COCO format, isn't it?

**Comment by SergiyShebotnov** [comment]
HF does not support datasets in the COCO format, they use their own HF format. You have to write your own COCO to HF converter for COCO segmentation datasets.
 I am yet to see public script loading instance segmentations in COCO format to HF format.

**Comment by NielsRogge** [comment]
I made [this notebook](https://colab.research.google.com/drive/1v1xzAT5mf12sQWD6Du2YC4y9NLXIYV9o?usp=sharing) to upload a COCO dataset to the hub.

The dataset lives here: https://huggingface.co/datasets/nielsr/coco-panoptic-val2017

**Comment by Kilikia123** [comment]
Please, can you show me you training and valid loss curve?
